src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Trimming.cs

```diff
@@ -0,0 +1,80 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Numerics;
+
+namespace System.Text
+{
+    public static partial class Ascii
+    {
+        /// <summary>
+        /// Trims all leading and trailing ASCII whitespaces from the buffer.
+        /// </summary>
+        /// <param name="value">The ASCII buffer.</param>
+        /// <returns>The Range of the untrimmed data.</returns>
+        public static Range Trim(ReadOnlySpan<byte> value) => TrimHelper(value, TrimType.Both);
+
+        /// <inheritdoc cref="Trim(ReadOnlySpan{byte})"/>
+        public static Range Trim(ReadOnlySpan<char> value) => TrimHelper(value, TrimType.Both);
+
+        /// <summary>
+        /// Trims all leading ASCII whitespaces from the buffer.
+        /// </summary>
+        /// <param name="value">The ASCII buffer.</param>
+        /// <returns>The Range of the untrimmed data.</returns>
+        public static Range TrimStart(ReadOnlySpan<byte> value) => TrimHelper(value, TrimType.Head);
+
+        /// <inheritdoc cref="TrimStart(ReadOnlySpan{byte})"/>
+        public static Range TrimStart(ReadOnlySpan<char> value) => TrimHelper(value, TrimType.Head);
+
+        /// <summary>
+        /// Trims all trailing ASCII whitespaces from the buffer.
+        /// </summary>
+        /// <param name="value">The ASCII buffer.</param>
+        /// <returns>The Range of the untrimmed data.</returns>
+        public static Range TrimEnd(ReadOnlySpan<byte> value) => TrimHelper(value, TrimType.Tail);
+
+        /// <inheritdoc cref="TrimEnd(ReadOnlySpan{byte})"/>
+        public static Range TrimEnd(ReadOnlySpan<char> value) => TrimHelper(value, TrimType.Tail);
+
+        private static Range TrimHelper<T>(ReadOnlySpan<T> value, TrimType trimType)
+            where T : unmanaged, IBinaryInteger<T>
+        {
+            const uint TrimMask =
+                (1u << (0x09 - 1))
+                | (1u << (0x0A - 1))
+                | (1u << (0x0B - 1))
+                | (1u << (0x0C - 1))
+                | (1u << (0x0D - 1))
+                | (1u << (0x20 - 1));
+
+            int start = 0;
+            if ((trimType & TrimType.Head) != 0)
+            {
+                for (; start < value.Length; start++)
+                {
+                    uint elementValue = uint.CreateTruncating(value[start]);
+                    if ((elementValue > 0x20) || ((TrimMask & (1u << ((int)elementValue - 1))) == 0))
```

Comment:
You could make this branch-free with:
```C#
        uint c = (ushort)(uint.CreateTruncating(value) - '\t');
        if ((int)((0xF8000100U << (short)c) & (c - 32)) >= 0)
```
See https://github.com/dotnet/runtime/blob/2480f0104160f8218d55568771b57c7f9c0d02f5/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs#L4599-L4628 for an explanation.

Reply:
If possible I would prefer to avoid adding further optimizations now, merge the API, add benchmarks to dotnet/performance and then let others tune it.